fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
-                if p.grad.requires_grad:
-                    raise RuntimeError("DistributedDataParallel only works with gradients that don't require grad")
+                if bucket.full():
 
-                p.grad.div_(world_size)
-                requests.append(dist.reduce(tensor=p.grad, dst=global_rank, group=group, async_op=True))  # type: ignore
+                    def unwrap() -> None:
+                        for flat in bucket.params:
+                            if dst_rank != self.global_rank:
+                                # this rank is not the owner, release the grad
+                                flat.param.grad = None
+                            else:
+                                # this rank is the owner, unroll the results
+                                assert flat.param.grad is not None
 
-        # Unroll the initial packed small gradients, as soon as possible
-        for future, rank in bucket_requests:
-            future.wait()
+                                flat.param.grad.data.copy_(
+                                    bucket.buffer[flat.start : flat.stop].view_as(flat.param.data), non_blocking=True
+                                )
 
-            if rank == self_rank:
-                i_bucketed = 0  # the number of tensors packed in the buffer
-                offset = 0
-                params = per_rank_params[rank]
-                buffer = buffers[rank]
+                        bucket.reset()
 
-                while i_bucketed < len(params) and offset + params[i_bucketed].numel() < buffer_size:
-                    end = offset + params[i_bucketed].numel()
-                    params[i_bucketed].grad.data.copy_(buffer[offset:end].view_as(params[i_bucketed]))  # type: ignore
-                    offset = end
-                    i_bucketed += 1
+                    bucket.buffer /= self.world_size
 
-        # Make sure that we're done with this device before moving on and cleaning the unused params
-        _ = list(map(lambda x: x.wait(), requests))
+                    optimizer.work_handles.append(
+                        Workhandle(
+                            handle=dist.reduce(
+                                tensor=bucket.buffer, dst=dst_rank, group=self.process_group, async_op=True,
+                            ),
+                            callback=unwrap,
+                        )
+                    )
+
+                    # If all the reduce operations have been called, add the gatekeeper
+                    if len(optimizer.work_handles) == optimizer._max_work_handles:
+                        gatekeeper()
 
-    def _sync_buffers(self) -> None:
+        return reduce_bucket if should_bucket else reduce_direct
+
+    def _setup_backward_hooks(self) -> None:
         """
-        Sync all the param buffers in between ranks.
-        TODO: Could be worth bucketing ?
+        Attach a reduce function to each grad-requiring parameter.
```

Comment:
What happens if I initialize a model and then later call `param.requires_grad=False` later after we've already wrapped the model? (I assume this usecase is legal)

Reply:
it's a good question, I had not thought of that. This case would not always work, and the reverse would never work:
- switch to requires_grad = False:
--  it would work if the param is not bucketed (it will not be part of the autograd and the hook will not be triggered)
-- it would not work if the param is bucketed, because it means that the bucket will never be full so never sent
- switch to requires_grad = True: will never work since this param will have no hook attached to it. I'm not sure how DDP handles that @mrshenli ?